File: android/domain/src/main/kotlin/com/jarvismobile/domain/port/Ports.kt
package com.jarvismobile.domain.port

import com.jarvismobile.domain.entity.AccountEntitlementSnapshot
import com.jarvismobile.domain.entity.ConfirmationRequest
import com.jarvismobile.domain.entity.PermissionType
import com.jarvismobile.domain.entity.UsageCost
import java.time.Instant

/**
 * Seams the pure [com.jarvismobile.domain] layer is built against instead of talking to any
 * platform API directly. Android implements these with real `ContextCompat`/Compose dialogs/
 * Retrofit calls; the backend implements the equivalent shapes server-side. See
 * ARCHITECTURE.md "Why a pure-Kotlin domain module".
 */
fun interface PermissionPort {
    suspend fun isGranted(permission: PermissionType): Boolean
}

fun interface EntitlementPort {
    suspend fun snapshot(accountId: String): AccountEntitlementSnapshot
}

fun interface UsagePort {
    /** Deducts [cost] for [skillId] and returns the account's new credit balance. */
    suspend fun charge(accountId: String, cost: UsageCost, skillId: String): Int
}

fun interface ConfirmationPort {
    suspend fun confirm(request: ConfirmationRequest): Boolean
}

fun interface ClockPort {
    fun now(): Instant
}

/** Default, real-clock implementation — swapped for a fixed clock in tests. */
object SystemClockPort : ClockPort {
    override fun now(): Instant = Instant.now()
}

File: backend/src/tooling/ports.ts
import type {
  AccountEntitlementSnapshot,
  PermissionType,
  RiskLevel,
  SkillExecutionContext,
  UsageCost,
} from "../domain/types.js";

/**
 * Platform seams the ToolPipeline is built against, mirroring android/domain/port/Ports.kt.
 * See ARCHITECTURE.md "Why a pure-Kotlin domain module" — the same reasoning applies here:
 * these interfaces keep the pipeline testable without a real database or auth provider.
 */
export interface PermissionPort {
  isGranted(accountId: string, permission: PermissionType): Promise<boolean>;
}

export interface EntitlementPort {
  snapshot(accountId: string): Promise<AccountEntitlementSnapshot>;
}

export interface UsagePort {
  /** Deducts `cost` for `skillId` and returns the account's new credit balance. */
  charge(accountId: string, cost: UsageCost, skillId: string): Promise<number>;
}

export interface ConfirmationRequest {
  skillId: string;
  summary: string;
  riskLevel: RiskLevel;
}

export interface ConfirmationPort {
  confirm(request: ConfirmationRequest, context: SkillExecutionContext): Promise<boolean>;
}

export interface ClockPort {
  now(): Date;
}

export const systemClockPort: ClockPort = { now: () => new Date() };
